contrib/haproxy/stream-processing-offload/haproxy_messages.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2025 Datadog, Inc.
+
+package streamprocessingoffload
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/DataDog/dd-trace-go/v2/instrumentation/appsec/proxy"
+	"github.com/negasus/haproxy-spoe-go/message"
+	"github.com/negasus/haproxy-spoe-go/request"
+
+	"github.com/DataDog/dd-trace-go/v2/ddtrace/ext"
+	"github.com/DataDog/dd-trace-go/v2/ddtrace/tracer"
+)
+
+var _ proxy.RequestHeaders = (*messageRequestHeaders)(nil)
+var _ proxy.ResponseHeaders = (*responseHeadersHAProxy)(nil)
+var _ proxy.HTTPBody = (*messageBody)(nil)
+var _ proxy.HTTPBody = (*messageBody)(nil)
+
+type messageRequestHeaders struct {
+	req     *request.Request
+	msg     *message.Message
+	hasBody bool
+}
+
+func (m *messageRequestHeaders) ExtractRequest(_ context.Context) (proxy.PseudoRequest, error) {
+	headers, err := parseHAProxyReqHdrsBin(getBytesArrayValue(m.msg, "headers"))
+	if err != nil {
+		return proxy.PseudoRequest{}, err
+	}
+
+	authority := headers.Get("Host")
+	method := getStringValue(m.msg, "method")
+	path := getStringValue(m.msg, "path")
+	https := getBoolValue(m.msg, "https")
+
+	if authority == "" || method == "" || path == "" {
+		return proxy.PseudoRequest{}, fmt.Errorf("missing required values in the http request SPOE message")
+	}
+
+	scheme := "http"
+	if https {
+		scheme = "https"
+	}
+
+	// Define if a body is present, based on Content-Length header
+	contentLength := headers.Get("Content-Length")
+	if contentLength != "" {
+		length, err := strconv.Atoi(contentLength)
+		if err != nil {
+			return proxy.PseudoRequest{}, fmt.Errorf("invalid Content-Length header: %v", err)
+		}
+		m.hasBody = length > 0
+	}
+
+	var remoteAddr string
+	remoteIp := getIPValue(m.msg, "ip")
+	if remoteIp != nil {
+		remotePort := strconv.Itoa(getIntValue(m.msg, "ip_port"))
+		remoteAddr = remoteIp.String() + ":" + remotePort
+	}
+
+	return proxy.PseudoRequest{
+		Method:     method,
+		Authority:  authority,
+		Path:       path,
+		Scheme:     scheme,
+		Headers:    headers,
+		RemoteAddr: remoteAddr,
+	}, nil
+}
+
+func (m *messageRequestHeaders) GetEndOfStream() bool {
+	return !m.hasBody
+}
+
+func (m *messageRequestHeaders) MessageType() proxy.MessageType {
+	return proxy.MessageTypeRequestHeaders
+}
+
+const componentNameHAProxySPOA = "haproxy-spoa"
+
+func (m *messageRequestHeaders) SpanOptions(_ context.Context) []tracer.StartSpanOption {
+	return []tracer.StartSpanOption{tracer.Tag(ext.Component, componentNameHAProxySPOA)}
+}
+
+type responseHeadersHAProxy struct {
+	msg     *message.Message
+	hasBody bool
+}
+
+func (m *responseHeadersHAProxy) ExtractResponse() (proxy.PseudoResponse, error) {
+	headers, err := parseHAProxyReqHdrsBin(getBytesArrayValue(m.msg, "headers"))
+	if err != nil {
+		return proxy.PseudoResponse{}, err
+	}
+
+	status := getIntValue(m.msg, "status_code")
+
+	// Set has body based on Content-Length header
+	contentLength := headers.Get("Content-Length")
+	if contentLength != "" {
```

Comment:
```suggestion
	if contentLength := headers.Get("Content-Length"); contentLength != "" {
```

Using the content length to know ahead of time if there will be a body is not exactly right I believe. Tons of streaming requests don't set the content-length header and still send a body. Either in the request or the response. cf. https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Content-Length

Although it can be used to know ahead of time if the body will be too big